File: arrays_and_strings/1_2.cs
// Given two strings write a method to decide if one is a permutation of another

using System;
using System.Collections;
using System.Collections.Generic;

namespace ArraysAndStrings {

    class Program {

        static void Main(string[] args) {

            Program program = new Program();

            var inputStrings = new List<(string, string)>
            {
                ("abc", "cba"),
                ("abc", "dct"), 
                ("ret", "re t"),
                ("gert", "ter")
            };
            
            foreach ((string string1, string string2) the_string in inputStrings)
            {
                //Console.WriteLine(the_string.string1);
                Console.WriteLine(program.is_permutation(the_string.string1, the_string.string2));
            }

        }

        bool is_permutation(string string1, string string2)
        {
            
            if(string1.Length != string2.Length)
            {
                return false;
            }

            int[] count_map = new int[128];

            for(int i=0; i < string1.Length; i++)
            {
                count_map[(int)string1[i]] += 1;
                Console.WriteLine($"Value: {count_map[(int)string1[i]]}");
            }

            for(int i=0; i < string1.Length; i++)
            {
                count_map[(int)string2[i]] -= 1;
                if(count_map[(int)string2[i]] < 0)
                {
                    return false;
                }
            }

            return true;
        }
    }
}

File: arrays_and_strings/1_2.py
# Given two strings write a method to decide if one is a permutation of another
                    #True
EXAMPLE_INPUTS = [('abc', 'cba'), ('abc', 'dct'), ('ret', 're t')]

def is_permutaion(s1, s2):

    # If the strings aren't the same size, they can't be permutations
    # (I managed to get this myself)
    if len(s1) != len(s2):
        return False
    
    for char in s1:
        if not char in s2:
            return False
    
    return True


# Her solution used a 128 size array where chars are inserted at their ASCII value
# eg 'a' would be insterted at 97 as ord('a') == 97
# Good space complexity
# My function above uses a built in function of Python, but how does the function work under the hood? 


for input in EXAMPLE_INPUTS:

    print(input, is_permutaion(input[0], input[1]))



# Questions I didn't ask:
# ASCII or Unicode
# Does Casing matter
# Whitespace??